core/mm3Types.js

```diff
@@ -65,10 +163,61 @@ CLASS({
       help: 'Regex pattern for property.'
     },
     {
+      name: 'minChars',
+      label: 'Minimum characters',
+      help: 'The minimum number of characters required.'
+    },
+    {
+      name: 'maxChars',
+      label: 'Maximum characters',
+      help: 'The maximum number of characters allowed.'
+    },
+    {
       name: 'prototag',
       label: 'Protobuf tag',
       required: false,
       help: 'The protobuf tag number for this field.'
+    },
+    {
+      name: 'validate',
+      lazyFactory: function() {
+        var prop = this; // this == the property
+        var ret = constantFn('');
+
+        var min = prop.minChars;
+        if ( typeof min !== 'string' || min ) { // ignore '' (falsey string)
+          min = parseInt(min);
+          ret = function(result) {
+            return result ||
+              ( this[prop.name].length < min ?
+                  prop.ERROR_BELOW_MIN_LENGTH.replaceValues(null, { min: min }) :
+                  ''
+              );
+          }.o(ret);
+        }
+        var max = prop.maxChars;
+        if ( typeof max !== 'string' || max ) { // ignore '' (falsey string)
+          max = parseInt(max);
+          ret = function(result) {
+            return result ||
+              ( this[prop.name].length > max ?
+                  prop.ERROR_ABOVE_MAX_LENGTH.replaceValues(null, { max: max }) :
+                  ''
+              );
+          }.o(ret);
+        }
+        var pattern = prop.pattern;
+        if ( pattern ) {
+          var testable = pattern.test ? pattern : new RegExp(pattern.toString(), 'i');
+          var errMsg = pattern.errorMessage ?
+            pattern.errorMessage() : prop.errorPatternMismatch;
+          ret = function(result) {
+            return result ||
+              ( ! testable.test(this[prop.name]) ? errMsg : '' );
+          }.o(ret);
+        }
+        return ret;
```

Comment:
All the arguments are result, which I suspect means that the dependencies are not being computed properly.

I think you'll have to set ret.dependencies = [prop.name] before returning, and in your validate property preSet, only compute the depenencies if not already set.

Be sure to test in something like FlightBooker that these validate functions are properly dynamic.